### PR TITLE
Transaction support

### PR DESCRIPTION
**Pull request by zachmu**
This change:
* Moves a bunch of transaction and other session management code out of Handler into Engine
* Introduces a new set of tests around transactions
* Fixes the error message for duplicate key violations
* Removes AsyncNode and related code
* Eliminates duplicate query parsing

### CODE REVIEW

File: engine.go [hunk 1]
```diff
@@ -143,101 +112,229 @@ func (e *Engine) AnalyzeQuery(
...
+			tdb, ok := database.(sql.TransactionDatabase)
+			if ok {
+				tx, err := tdb.StartTransaction(ctx)
+				if err != nil {
+					return nil, nil, err
+				}
+				ctx.SetTransaction(tx)
+			}
+		}
+	}
```
Comment (andrew-wm-arthur) [inline engine.go:215]: refactor suggestion: wrap the `transactionCommittingIter` here with the `transactionDatabase` embedded

File: enginetest/testdata.go [hunk 2]
```diff
@@ -39,428 +39,454 @@ func includeTable(includedTables []string, tableName string) bool {
 	return false
 }
 
+func wrapInTransaction(t *testing.T, db sql.Database, harness Harness, fn func()) {
+	if tdb, ok := db.(sql.TransactionDatabase); ok {
```
Comment (andrew-wm-arthur) [inline enginetest/testdata.go:43]: we shouldn't run this if it's not a `TransactionalDatabase`, we should `t.skip()` or panic

**Review by andrew-wm-arthur — APPROVED** [review]
LGTM!

File: engine.go (hunk 1 above)
Comment (zachmu) [reply engine.go:215]: Good call, this revealed some other grossness

File: enginetest/testdata.go (hunk 2 above)
Comment (zachmu) [reply enginetest/testdata.go:43]: You kinda have to... it's a no-op for databases that don't support transactions, but required for those that do

I'll add a comment

File: enginetest/testdata.go [hunk 3]
```diff
@@ -39,428 +39,454 @@ func includeTable(includedTables []string, tableName string) bool {
...
-				sql.NewRow(0, 0, 1, 2, 3, 4),
-				sql.NewRow(1, 10, 11, 12, 13, 14),
-				sql.NewRow(2, 20, 21, 22, 23, 24),
-				sql.NewRow(3, 30, 31, 32, 33, 34))
-		} else {
-			t.Logf("Warning: could not create table %s: %s", "one_pk", err)
+	// This is a bit odd, but because this setup doesn't interact with the engine.Query path, we need to do transaction
+	// management here, instead. If we don't, then any Query-based setup will wipe out our work by starting a new
+	// transaction without committing the work done so far.
+	// The secondary foo database doesn't have this problem because we don't mix and match query and non-query setup
```
[the remark this answers is not in the document]
Comment (zachmu) [reply enginetest/testdata.go:69]: The issue is that then we couldn't run the engine tests on read only database implementations, so it's worth some amount of futzing to support

[the remark this answers is not in the document]
Comment (zachmu) [reply enginetest/testdata.go:71]: Actually now that I'm saying that, it's already not true, they have to implement sql.InsertableTable at a minimum already

**Pull request merged**